problem-gson/src/main/java/org/zalando/problem/ProblemAdapterFactory.java

```diff
@@ -0,0 +1,150 @@
+                                  Map<URI, Class<? extends Problem>> subtypes) {
+        this.stackTraces = stackTraces;
+        this.statusAdapter = statusAdapter;
+        this.subtypes = Collections.unmodifiableMap(subtypes);
+    }
+
+    @SafeVarargs
+    private static <E extends Enum & StatusType> Map<Integer, StatusType> buildIndex(
+            final Class<? extends E>... types) {
+        final Map<Integer, StatusType> index = new HashMap<>();
+
+        for (final Class<? extends E> type : types) {
+            for (final E status : type.getEnumConstants()) {
+                if (index.containsKey(status.getStatusCode())) {
+                    throw new IllegalArgumentException("Duplicate status codes are not allowed");
+                }
+                index.put(status.getStatusCode(), status);
+            }
+        }
+
+        return Collections.unmodifiableMap(index);
+    }
+
+    public ProblemAdapterFactory withStacktraces() {
+        return withStacktraces(true);
+    }
+
+    public ProblemAdapterFactory withStacktraces(final boolean stackTraces) {
+        return new ProblemAdapterFactory(stackTraces, this.statusAdapter, this.subtypes);
+    }
+
+    @CheckReturnValue
+    public <T extends Problem> ProblemAdapterFactory registerSubtype(Class<T> clazz, URI type) {
+        Objects.requireNonNull(clazz);
+        Objects.requireNonNull(type);
+        if (subtypes.containsKey(type) || subtypes.containsValue(clazz)) {
+            throw new IllegalArgumentException("class & type must be unique");
+        }
+        HashMap<URI, Class<? extends Problem>> map = new HashMap<>(subtypes);
+        map.put(type, clazz);
+        return new ProblemAdapterFactory(stackTraces, statusAdapter, map);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
+        Class<? super T> rawType = type.getRawType();
+
+        if (StatusType.class.isAssignableFrom(rawType)) {
+            return (TypeAdapter<T>) statusAdapter;
+        }
+
+        if (!Problem.class.isAssignableFrom(rawType)) {
+            return null;
+        }
+
+        DefaultProblemAdapter defaultAdapter = new DefaultProblemAdapter(gson, stackTraces);
+        return new TypeAdapter<T>() {
+            @Override
+            public void write(JsonWriter out, T value) throws IOException {
+                if (value instanceof DefaultProblem) {
+                    defaultAdapter.write(out, (DefaultProblem) value);
+                    return;
+                }
+
+                TypeAdapter<T> delegate = (TypeAdapter<T>) gson.getDelegateAdapter(ProblemAdapterFactory.this, TypeToken.get(value.getClass()));
+                JsonElement element = delegate.toJsonTree(value);
+                JsonObject jsonObject = element.getAsJsonObject();
+                Optional.ofNullable(jsonObject.remove("stackTrace"))
```

Comment:
Is there no cleaner way?

Reply:
Aah this, I am also not very happy with this part of code. I'll keep working and update if I come up with something better.

Reply:
@whiskeysierra This particular piece handles the case when a Custom Problem is given for serialization, by delegating the serialization to Gson first and then transform certain fields to match expected output. 

Currently I'm transforming `stackTrace` field because:
- first the field name does not conforms to all lowercase **stacktrace** expected in the output, and
- second Gson does not serialize `StackTraceElement[] getStackTrace()` to `String[]`, rather it spits something like:
    ```json
    [
         {
            "declaringClass": "org.zalando.problem.ProblemBuilder",
            "methodName": "build",
            "fileName": "ProblemBuilder.java",
            "lineNumber": 83
          }
    ]
    ```
- third, I can't attach a custom adapter for StackTraceElement[] via factory since it might change Gson's behavior across application if someone is using a shared instance of Gson (I don't know how plausible is this).

Also, a side-effect that I observed just now as a result of delegating to Gson is that it's serializing fields like `"suppressedExceptions": []` (since non-null) as well. Is this reasonable for Custom Problems?

Reply:
>     * third, I can't attach a custom adapter for StackTraceElement[] via factory since it might change Gson's behavior across application if someone is using a shared instance of Gson (I don't know how plausible is this).

Is the same true for status and uri?

Reply:
> Also, a side-effect that I observed just now as a result of delegating to Gson is that it's serializing extra fields like `"suppressedExceptions": [] - from Throwable` (since non-null) as well. Is this reasonable for Custom Problems?

Nope. The jackson module ignores all properties that are inherited from Throwable, unless explicitly allowed.

Reply:
See e.g. https://github.com/zalando/problem/blob/master/jackson-datatype-problem/src/main/java/org/zalando/problem/ExceptionalMixin.java#L29

Reply:
> Is the same true for status and uri?

Yes, although current implementation will only attach an adapter for StatusType and not for URI. Also this made me realize that `type: URI` field is not properly serialized for Custom Problem when its value is `null` 😟.

Reply:
> > Also, a side-effect that I observed just now as a result of delegating to Gson is that it's serializing extra fields like `"suppressedExceptions": [] - from Throwable` (since non-null) as well. Is this reasonable for Custom Problems?
> 
> Nope. The jackson module ignores all properties that are inherited from Throwable, unless explicitly allowed.

Alright, I better revisit this and come up with correct implementation.

Reply:
> See e.g. https://github.com/zalando/problem/blob/master/jackson-datatype-problem/src/main/java/org/zalando/problem/ExceptionalMixin.java#L29

I did refer this and tried to cover possible cases, but the hiccup is that there's no Jackson Mixin-like support in Gson. I guess I will have to mimic this behavior for Gson to make this work.

Reply:
> Is there no cleaner way?

@whiskeysierra in the latest commit, I've extracted the custom problem serialization logic in a new adapter, still using previous logic but with care to ignore extra fields from `Throwable`.

Also appended JSON field count assertions to all test cases to ensure no extra field gets serialized besides the expected ones.